### DISCUSSION TITLE
Equivalent of CompositionTarget.Rendering event?

### DISCUSSION THREAD

**Discussion by rmanosuthi** [AvaloniaUI/Avalonia#9337] [Q&A] [answered]
Hello,

I've dug around Avalonia's docs and can't seem to find a suitable substitute for this WPF event. I need to do some stuffs every time the screen is rendered.

**Comment by grokys** [comment]
Hey @mpipo - we don't actually have an equivalent of that at the moment. Our render/animation loop is currently a mess.

https://github.com/AvaloniaUI/Avalonia/pull/1715 aims to fix that, and before we finish it, we will add the equivalent of `CompositonTarget.Rendering`.

For the moment though, your best bet is probably to subscribe to one of the timers in https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Animation/Timing.cs, such as [the one returned by this method](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Animation/Timing.cs#L109). 

Note that when #1715 is merged however this API will change.

**Comment by jhimes144** [comment]
@grokys I see that pr was merged a while ago. I'm a new comer to Avalonia but very familiar with WPF. What is the equivalent now of `CompositionTarget.Rendering`? I have a simple test where I have 400 rectangles all changing random colors. I want the changes to happen on the ui thread so that for every tick new colors are drawn. This works fast in WPF with CompositionTarget.Rendering at 60fps but I've tried various timers in Avalonia and I get horrible performance.

**Comment by kekekeks** [comment]
@jhimes144 please check if "SkCanvas" page from the [RenderDemo](https://github.com/AvaloniaUI/Avalonia/tree/master/samples/RenderDemo) project has acceptable performance

**Comment by grokys** [comment]
If you don't want to go straight to Skia, I think the closest to `CompositionTarget.Rendering` would be to get hold of the [`IRenderLoop`](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Visuals/Rendering/IRenderLoop.cs) from `AvaloniaLocator` and register an `IRenderLoopTask`.

I know we said we'd add the equivalent of `CompositonTarget.Rendering` but seems we didn't...

**Comment by jhimes144** [comment]
@grokys So I tried that and also tried creating an event handle off IRenderTimer Tick directly. Whats interesting is that when I do any UI related tasks I get an invalid thread exception. So I got curious and tried this line `Dispatcher.UIThread.VerifyAccess();` and it threw an exception. I'm a bit confused shouldn't that event be on the UI thread?

**Comment by kekekeks** [comment]
IRenderLoop ticks on the render thread.

**Comment by grokys** [comment]
`IRenderLoopTask` has two methods: `Render` and `Update`. `Render` is called on the render thread, but `Update` will be called on the UI thread when `NeedsUpdate` is set to true.

**Comment by jhimes144** [comment]
Coming back to this 2 years later.

Wrote the same test in Avalonia using the `IRenderLoopTask` (correctly?). This yields roughly ~20fps performance in release. Avalonia utilizes 100% of one cpu core running at 4.9GHz. WPF fairs way better with a handle of `CompositonTarget.Rendering`.

I'm interested in building an application that shows data in real time polling without needing to fall back to drawing bitmaps. Is there anyway I can improve the performance? Also Dispatching the updates of course is way worst in terms of performance.

ALSO SOMETHING VERY STRANGE TO NOTE:

When running in visual studio's preview, the test runs at 32fps??!

Another note:
Ran another test using SkiaSharp canvas. Performance was very good, downside is of course loosing out on the Avalonia platform to draw UI.

`public partial class MainWindow : Window
    {
        public MainWindow()
        {
            InitializeComponent();

            for (int i = 0; i < 400; i++)
            {
                var rect = new Rectangle();
                rect.Width = 20;
                rect.Height = 20;
                rect.Stroke = null;
                rect.Fill = new SolidColorBrush(Colors.White);

                TestPanel.Children.Add(rect);
            }

            Avalonia.AvaloniaLocator.Current.GetService<IRenderLoop>().Add(new RenderLoopTaskTest
            {
                TestPanel = TestPanel,
                Fps = FPS
            });
        }
    }

    public class RenderLoopTaskTest : IRenderLoopTask
    {
        public bool NeedsUpdate => true;

        public WrapPanel TestPanel { get; set; }
        public TextBlock Fps { get; set; }

        private TimeSpan _lastFrame;

        public void Render()
        {
            
        }

        public void Update(TimeSpan time)
        {
            if (time - _lastFrame > TimeSpan.FromMilliseconds(16))
            {
                var random = new Random();

                foreach (var rect in TestPanel.Children.Cast<Rectangle>())
                {
                    var r = random.Next(0, 255);
                    var g = random.Next(0, 255);
                    var b = random.Next(0, 255);
                    var s = (SolidColorBrush)rect.Fill;
                    s.Color = new Color(255, (byte)r, (byte)g, (byte)b);
                }

                Fps.Text = 1000d / (time - _lastFrame).TotalMilliseconds + "fps";
                _lastFrame = time;
            }
        }
    }`

**Comment by maxkatz6** [comment, marked as answer]
> Wrote the same test in Avalonia using the IRenderLoopTask (correctly?)

I wonder, why have you chosen it? I don't think we have mentioned this interface as a solution. If we did, I would correct it.

Normally it's enough to override Render method. This method runs **on UI thread** and drawing commands are deferred.
Alternatively, for better performance you can use ICustomDrawOperation. This operation is executed **on Render thread**. You can create new instance of the draw operation in your Update method (or edit previous, though carefully with threads) and call draw operations in this draw operation implementation. 

As mentioned above, we have skia rendering demo made in this way. If you don't need special skia APIs, you can just call draw methods on IDrawingContextImpl directly.
https://github.com/AvaloniaUI/Avalonia/blob/master/samples/RenderDemo/Pages/CustomSkiaPage.cs

**Reply by jgcodes2020** [reply]
For controls exposing a CompositionSurface, what's the equivalent (given that CompositionVisual entirely bypasses Render())?

**Reply by kekekeks** [reply]
Compositor.RequestCompositionUpdate - callback is called before the next composition batch is sent to the render thread. That happens strictly after layout, so any changes to the layout won't be applied until the next frame.

TopLevel.RequestAnimationFrame - callback is called during the next animation pass, that happens before layout and is throttled to the compositor update rate. You can affect layout there.

So for surfaces, you probably want to use the first one.

**Comment by jhimes144** [comment]
@maxkatz6 I chose the IRenderLoopTask because it was suggested as a solution in this thread 2 years ago.

I wrote a test with a user control's render method overridden with 400 rectangles changing random colors and I got 100fps performance. This is probably good enough for me. Thanks for the help.

**Comment by timunie** [comment]
Converted this into Q&A as this is neither a bug nor a feature request. If you disagree please let us know.
